### ISSUE TITLE
query.js line 146 throwing an exception when error message gets truncated

### ISSUE THREAD

**Issue by radoslavpetranov** [sequelize/sequelize#2965]
I'm hitting a foreign key constraint when trying to delete some records and this is the error that I end up with 

 { 
    [Error: ER_ROW_IS_REFERENCED_: Cannot delete or update a parent row: a foreign key constraint fails (`aaaaaaa_aaaaaa_aaaaaaa`.`report_templated_text_variable_entry`, CONSTRAINT `report_templated_text_variable_entry_ibfk_3` FOREIGN KEY (`template_var_id`) REFERENCES `report_templated_text_vari)]
  code: 'ER_ROW_IS_REFERENCED_',
  errno: 1451,
  sqlState: '23000',
  index: 0,
 sql: 'DELETE FROM`report_templated_text_variable_xref`WHERE`id` IN (524)' }

This is the full stack of the error

TypeError: Cannot read property '3' of null
    at module.exports.Query.formatError (c:\whatever\node_modules\sequelize\lib\dialects\mysql\query.js:146:23)
    at Query._callback (c:\whatever\node_modules\sequelize\lib\dialects\mysql\query.js:38:23)
    at Query.Sequence.end (c:\whatever\node_modules\mysql\lib\protocol\sequences\Sequence.js:96:24)
    at Query.ErrorPacket (c:\whatever\node_modules\mysql\lib\protocol\sequences\Query.js:93:8)
    at Protocol._parsePacket (c:\whatever\node_modules\mysql\lib\protocol\Protocol.js:271:23)
    at Parser.write (c:\whatever\node_modules\mysql\lib\protocol\Parser.js:77:12)
    at Protocol.write (c:\whatever\node_modules\mysql\lib\protocol\Protocol.js:39:16)
    at Socket.<anonymous> (c:\whatever\node_modules\mysql\lib\Connection.js:82:28)
    at Socket.EventEmitter.emit (events.js:95:17)
    at Socket.<anonymous> (_stream_readable.js:746:14)

What seems to be happening is that the error message above gets truncated (notice how the table name is never completed "REFERENCES `report_templated_text_vari)") so the match fails and index: match[3] ends throwing an exception. 

Not exactly a show stopper but I thought I'd share in case you guys care about this.

**Comment by mickhansen** [comment]
We definitely care :)
Can you show the sequelize models and associations that generate this constraint?

**Comment by radoslavpetranov** [comment]
Well, I submitted this as a bug in the error processing code only (i.e. handling the truncated error) thinking that I'd figure out what I'm doing wrong but since I still haven't figured it out here are the models + associations - let's see if I can get this to work automatically as opposed to parsing the associations myself and deleting everything one by one:

this.dbObjects.ReportTemplatedTextVariable = this.sequelizeLib.define('report_templated_text_variable_xref', {
	id: {
		type: Sequelize.INTEGER,
		primaryKey: 'true',
		autoIncrement: true
	},
        // more unrelated columns here ..
});

this.dbObjects.ReportTemplatedTextVariableEntry = this.sequelizeLib.define('report_templated_text_variable_entry', {
	id: {
		type: Sequelize.INTEGER,
		primaryKey: 'true',
		autoIncrement: true
	},
	// .. some other columns here
});

// Then in the associations I've got this
this.dbObjects.ReportTemplatedTextVariable.hasMany(this.dbObjects.ReportTemplatedTextVariableEntry, {as:'entries', foreignKey: 'template_var_id', foreignKeyConstraint: true, onDelete : 'cascade'});

And in mySQL we've got this for the report_templated_text_variable_entry table

...
KEY `template_var_id` (`template_var_id`),
...
CONSTRAINT `report_templated_text_variable_entry_ibfk_3` FOREIGN KEY (`template_var_id`) REFERENCES `report_templated_text_variable_xref` (`id`)
...

So basically the error happens when I run

this.dbObjects.ReportTemplatedTextVariable.destroy({where : {id : id}, transaction: t }) ..

the onDelete:cascade doesn't seem to fire up so the entries for that variable are still there and I hit the fk constraint.

Cheers and thanks for the help

**Comment by janmeier** [comment]
This was an issue with lodash.cloneDeep (used in the upsert code) which does not preserve `Buffer.isBuffer`

**Comment by ssteffl** [comment]
i ran into a similar issue with ER_DUP_ENTRY, with the error message being empty and Sequelize causing an error to be thrown in Query.formatError. It seems to be a bug coming from a lower layer.

https://github.com/felixge/node-mysql/issues/1397
